### PR TITLE
CDAP-14932. Salesforce Batch Source to fetch objects via SOQL. Integraion tests

### PR DESCRIPTION
**Pull request by aonischuk**
parallel='none', perCoreThreadCount=true, threadCount=2, useUnlimitedThreads=false
Running co.cask.hydrator.salesforce.etl.SalesforceBatchSourceETLTest
Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 57.013 sec
testTypesConversion(co.cask.hydrator.salesforce.etl.SalesforceBatchSourceETLTest)  Time elapsed: 13.176 sec
testSOQLSyntaxError(co.cask.hydrator.salesforce.etl.SalesforceBatchSourceETLTest)  Time elapsed: 1.108 sec
testEmptyQueryResponse(co.cask.hydrator.salesforce.etl.SalesforceBatchSourceETLTest)  Time elapsed: 9.163 sec
testWhereClause(co.cask.hydrator.salesforce.etl.SalesforceBatchSourceETLTest)  Time elapsed: 10.395 sec
testValuesReturned(co.cask.hydrator.salesforce.etl.SalesforceBatchSourceETLTest)  Time elapsed: 9.116 sec
testLikeClause(co.cask.hydrator.salesforce.etl.SalesforceBatchSourceETLTest)  Time elapsed: 9.41 sec
parallel='none', perCoreThreadCount=true, threadCount=2, useUnlimitedThreads=false
Running co.cask.hydrator.salesforce.SalesforceSchemaUtilTest
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.279 sec
testGetSchemaWithFields(co.cask.hydrator.salesforce.SalesforceSchemaUtilTest)  Time elapsed: 0.222 sec
testSyntaxError(co.cask.hydrator.salesforce.SalesforceSchemaUtilTest)  Time elapsed: 0.056 sec
parallel='none', perCoreThreadCount=true, threadCount=2, useUnlimitedThreads=false
Running co.cask.hydrator.salesforce.plugin.source.batch.SalesforceRecordReaderTest
Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.404 sec
testTypes(co.cask.hydrator.salesforce.plugin.source.batch.SalesforceRecordReaderTest)  Time elapsed: 0.4 sec
testUTF8InKeysAndValues(co.cask.hydrator.salesforce.plugin.source.batch.SalesforceRecordReaderTest)  Time elapsed: 0.001 sec
testLineBreakAndCommaInCSV(co.cask.hydrator.salesforce.plugin.source.batch.SalesforceRecordReaderTest)  Time elapsed: 0 sec
testKeysAndValuesDifferentNumber(co.cask.hydrator.salesforce.plugin.source.batch.SalesforceRecordReaderTest)  Time elapsed: 0.001 sec
testInvalidCSV(co.cask.hydrator.salesforce.plugin.source.batch.SalesforceRecordReaderTest)  Time elapsed: 0 sec
testEmptyCSVResponse(co.cask.hydrator.salesforce.plugin.source.batch.SalesforceRecordReaderTest)  Time elapsed: 0 sec
parallel='none', perCoreThreadCount=true, threadCount=2, useUnlimitedThreads=false
Running co.cask.hydrator.salesforce.parser.SalesforceQueryParserTest
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.077 sec
testGetsObjectFromQuery(co.cask.hydrator.salesforce.parser.SalesforceQueryParserTest)  Time elapsed: 0.071 sec
testGetFieldsFromQuery(co.cask.hydrator.salesforce.parser.SalesforceQueryParserTest)  Time elapsed: 0.005 sec

Results :

Tests run: 16, Failures: 0, Errors: 0, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:15 min
[INFO] Finished at: 2019-04-01T23:11:49+03:00
[INFO] ------------------------------------------------------------------------

### CODE REVIEW

File: src/test/java/co/cask/hydrator/salesforce/etl/BaseSalesforceBatchSourceETLTest.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+public abstract class BaseSalesforceBatchSourceETLTest extends HydratorTestBase {
+
+  @ClassRule
+  public static final TestConfiguration CONFIG = new TestConfiguration("explore.enabled", false);
+
+  private static final ArtifactSummary APP_ARTIFACT = new ArtifactSummary("data-pipeline", "3.2.0");
+
+  // TODO: move to properties
+  // TODO: ignore if not set
+  private static final String CLIENT_ID = System.getProperty("sf.test.clientId");
```
Comment (aielchiieva) [inline src/test/java/co/cask/hydrator/salesforce/etl/BaseSalesforceBatchSourceETLTest.java]:
```suggestion
  private static final String CLIENT_ID = System.getProperty("salesforce.test.clientId");
```

Comment (aielchiieva) [reply src/test/java/co/cask/hydrator/salesforce/etl/BaseSalesforceBatchSourceETLTest.java]: I suggest to replace all prefixes "sf."with "salesforce."

File: src/test/java/co/cask/hydrator/salesforce/etl/BaseSalesforceBatchSourceETLTest.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+  private static final String PASSWORD = System.getProperty("sf.test.password");
+  private static final String LOGIN_URL = System.getProperty("sf.test.loginUrl",
+                                                            "https://login.salesforce.com/services/oauth2/token");
+
+  @BeforeClass
+  public static void setupTestClass() throws Exception {
+    try {
+      Assume.assumeNotNull(CLIENT_ID, CLIENT_SECRET, USERNAME, PASSWORD, LOGIN_URL);
+    } catch (AssumptionViolatedException e) {
+      System.out.println("WARNING: ETL tests are skipped. Please find the instructions on enabling it at" +
```
Comment (aielchiieva) [inline src/test/java/co/cask/hydrator/salesforce/etl/BaseSalesforceBatchSourceETLTest.java]: Add message to runtime exception and remove System.out.println

File: src/test/java/co/cask/hydrator/salesforce/etl/BaseSalesforceBatchSourceETLTest.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    // this will make our plugins available.
+
+    addPluginArtifact(NamespaceId.DEFAULT.artifact("example-plugins", "1.0.0"),
+                      parentArtifact,
+                      SalesforceBatchSource.class);
+  }
+
+  protected List<StructuredRecord> getResultsBySOQLQuery(String query) throws Exception {
+    ImmutableMap.Builder<String, String> propsBulder = new ImmutableMap.Builder<String, String>()
+      .put("referenceName", "SalesforceBulk" + UUID.randomUUID().toString())
```
Comment (aielchiieva) [inline src/test/java/co/cask/hydrator/salesforce/etl/BaseSalesforceBatchSourceETLTest.java]: better to set static name to reference. E.g. SalesforceBulk-input

File: src/test/java/co/cask/hydrator/salesforce/etl/BaseSalesforceBatchSourceETLTest.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+  private List<StructuredRecord> getPipelineResults(Map<String, String> sourceProperties) throws Exception {
+    ETLStage source = new ETLStage("SalesforceReader", new ETLPlugin("SalesforceBulk",
+                                                                     BatchSource.PLUGIN_TYPE,
+                                                                     sourceProperties, null));
+
+    String outputDatasetName = "output-batchsourcetest" + UUID.randomUUID().toString();
+    ETLStage sink = new ETLStage("sink", MockSink.getPlugin(outputDatasetName));
+
+    ETLBatchConfig etlConfig = ETLBatchConfig.builder("* * * * *")
```
Comment (aielchiieva) [inline src/test/java/co/cask/hydrator/salesforce/etl/BaseSalesforceBatchSourceETLTest.java]:
```suggestion
    ETLBatchConfig etlConfig = ETLBatchConfig.builder()
```

File: src/test/java/co/cask/hydrator/salesforce/etl/BaseSalesforceBatchSourceETLTest.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+      .addStage(source)
+      .addStage(sink)
+      .addConnection(source.getName(), sink.getName())
+      .build();
+
+    ApplicationId pipelineId = NamespaceId.DEFAULT.app("testPipeline" + UUID.randomUUID().toString());
+    ApplicationManager appManager = deployApplication(pipelineId, new AppRequest<>(APP_ARTIFACT, etlConfig));
+
+    WorkflowManager workflowManager = appManager.getWorkflowManager(SmartWorkflow.NAME);
+    workflowManager.start();
```
Comment (aielchiieva) [inline src/test/java/co/cask/hydrator/salesforce/etl/BaseSalesforceBatchSourceETLTest.java]:
```suggestion

```

File: src/test/java/co/cask/hydrator/salesforce/etl/BaseSalesforceBatchSourceETLTest.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    workflowManager.waitForRuns(ProgramRunStatus.COMPLETED, 1, 5, TimeUnit.MINUTES);
```
Comment (aielchiieva) [inline src/test/java/co/cask/hydrator/salesforce/etl/BaseSalesforceBatchSourceETLTest.java]:
```suggestion
    workflowManager.startAndWaitForRun(ProgramRunStatus.COMPLETED, 1, 5, TimeUnit.MINUTES);
```

File: src/test/java/co/cask/hydrator/salesforce/etl/BaseSalesforceBatchSourceETLTest.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    String outputDatasetName = "output-batchsourcetest" + UUID.randomUUID().toString();
+    ETLStage sink = new ETLStage("sink", MockSink.getPlugin(outputDatasetName));
+
+    ETLBatchConfig etlConfig = ETLBatchConfig.builder("* * * * *")
+      .addStage(source)
+      .addStage(sink)
+      .addConnection(source.getName(), sink.getName())
+      .build();
+
+    ApplicationId pipelineId = NamespaceId.DEFAULT.app("testPipeline" + UUID.randomUUID().toString());
```
Comment (aielchiieva) [inline src/test/java/co/cask/hydrator/salesforce/etl/BaseSalesforceBatchSourceETLTest.java]: create applicationId based on plugin name + test name

File: src/test/java/co/cask/hydrator/salesforce/etl/SalesforceBatchSourceETLTest.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+                                            Schema.Field.of("LastModifiedDate",
+                                                            Schema.of(Schema.LogicalType.TIMESTAMP_MILLIS))
+    );
+
+    Assert.assertEquals(expectedSchema, record.getSchema());
+  }
+
+  @Test
+  public void testValuesReturned() throws Exception {
+    String query = "SELECT OwnerId, Name, IsDeleted FROM Account"; // select only non-nillable fields
```
Comment (aielchiieva) [inline src/test/java/co/cask/hydrator/salesforce/etl/SalesforceBatchSourceETLTest.java]: You should insert data before test then fetch data by id or by expected field values to be sure that inserted data fetched properly. Please add datasets for each case and test all supported Salesforce data types.

File: src/test/java/co/cask/hydrator/salesforce/etl/BaseSalesforceBatchSourceETLTest.java [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Methods to run ETL with Salesforce Bulk plugin as source, and a mock plugin as a sink.
+ *
+ * By default all tests will be skipped, since Salesforce credentials are needed.
+ *
+ * Instructions to enable the tests:
+ * 1. Create/use existing Salesforce account
+ * 2. Tables need to have some records. Defaults setup by Salesforce during account creation are enough.
```
Comment (aielchiieva) [inline src/test/java/co/cask/hydrator/salesforce/etl/BaseSalesforceBatchSourceETLTest.java]: Integration test must create new sObject, insert data and then remove in the end for integration test run but not rely on the data to be present in Salesforce already.

File: src/test/java/co/cask/hydrator/salesforce/etl/BaseSalesforceBatchSourceETLTest.java [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+ *
+ */
+public abstract class BaseSalesforceBatchSourceETLTest extends HydratorTestBase {
+
+  @ClassRule
+  public static final TestConfiguration CONFIG = new TestConfiguration("explore.enabled", false);
+
+  private static final ArtifactSummary APP_ARTIFACT = new ArtifactSummary("data-pipeline", "3.2.0");
+
+  // TODO: move to properties
```
Comment (aielchiieva) [inline src/test/java/co/cask/hydrator/salesforce/etl/BaseSalesforceBatchSourceETLTest.java]: remove todo's

File: src/test/java/co/cask/hydrator/salesforce/etl/BaseSalesforceBatchSourceETLTest.java (hunk 2 above)
Comment (aonischuk) [reply src/test/java/co/cask/hydrator/salesforce/etl/BaseSalesforceBatchSourceETLTest.java]: We need AssumptionViolatedException so the tests are skipped. Not runtime.

Comment (aielchiieva) [reply src/test/java/co/cask/hydrator/salesforce/etl/BaseSalesforceBatchSourceETLTest.java]: The main point here is to remove `System.out.println`.  You can replace it with logger.

**Review by aielchiieva — APPROVED** [review]
+1

**Pull request merged**